apps/workspace-host/src/interface.ts

```diff
@@ -33,6 +33,43 @@ import { config, loadConfig } from './lib/config.js';
 import { parseDockerLogs } from './lib/docker.js';
 import { REPOSITORY_ROOT_PATH, APP_ROOT_PATH } from './lib/paths.js';
 import * as socketServer from './lib/socket-server.js';
+let server: http.Server | undefined;
+
+interface WorkspaceServerSettings {
+  instance_id?: string;
+  hostname?: string;
+  server_to_container_hostname?: string;
+  port?: number;
+}
+
+interface WorkspaceSettings {
+  workspace_image: string;
+  workspace_port: number;
+  workspace_home: string;
+  workspace_graded_files: string[];
+  workspace_args: string;
+  workspace_enable_networking: boolean;
+  workspace_environment: string[];
+}
+
+/*
+These types are somewhat duplicated from WorkspaceSchema in apps/prairielearn.
+TODO: break out types into a shared package.
+*/
+interface Workspace {
+  local_name: string;
+  launch_uuid: string;
+  remote_name: string;
+  launch_port?: number;
+  id: string | number;
+  version: string;
+  course_id: string;
+  institution_id: string;
+  settings?: WorkspaceSettings;
+  container?: Docker.Container;
```

Comment:
I'm reasonably confident that you can restructure code such that these can be made required properties. You'll have to change the initialization process in `initSequence` a little but, but I think it's doable!

Reply:
I did a potential restructure -- mainly by just splitting out settings and container into seperate variables. Due to data dependencies, it seems nontrivial to do this without some "typescript magic", but I may be missing something.